cmd/kubemark/hollow-node.go

```diff
@@ -173,60 +160,92 @@ func run(config *hollowNodeConfig) {
-		config.TenantServers[0] = clientConfigs.GetConfig().Host
+		config.TenantServers[0] = config.KubeconfigPath
 	}
 	if config.ResourceServer == "" {
 		klog.V(3).Infof("Resource is not set. Default to clientConfig setting")
-		config.ResourceServer = clientConfigs.GetConfig().Host
+		config.ResourceServer = config.KubeconfigPath
 	}
 
 	// initialize the kubeclient manager
 	kubeclientmanager.NewKubeClientManager()
 
+	// TODO: set this from the hollow node env variable
+	enableInSecurePorts := false
+
 	numberTenantPartitions := len(config.TenantServers)
 	clients := make([]clientset.Interface, numberTenantPartitions)
-	for i := 0; i < len(clients); i++ {
-		tenantCfg := restclient.CopyConfigs(clientConfigs)
-		for _, cfg := range tenantCfg.GetAllConfigs() {
-			cfg.Host = config.TenantServers[i]
-		}
-		clientFromConfig, err := clientset.NewForConfig(tenantCfg)
-		if err != nil {
-			klog.Fatalf("Failed to create a ClientSet: %v. Exiting.", err)
+	for i := 0; i < numberTenantPartitions; i++ {
+		if enableInSecurePorts == false {
+			kubeconfigFile := config.TenantServers[i]
+			klog.V(2).Infof("create client config from file: %s", kubeconfigFile)
+			cfg, err := clientutil.CreateClientConfigFromKubeconfigFileAndSetQps(kubeconfigFile, 10, 20, config.ContentType)
+			if err != nil {
+				klog.Fatalf("Failed to create a client config: %v. Exiting.", err)
+			}
+
+			clientFromConfig, err := clientset.NewForConfig(cfg)
+			if err != nil {
+				klog.Fatalf("Failed to create a ClientSet: %v. Exiting.", err)
+			}
+			clients[i] = clientFromConfig
+		} else {
+			tenantCfg := restclient.CopyConfigs(clientConfigs)
+			for _, cfg := range tenantCfg.GetAllConfigs() {
+				cfg.Host = config.TenantServers[i]
+			}
+			clientFromConfig, err := clientset.NewForConfig(tenantCfg)
+			if err != nil {
+				klog.Fatalf("Failed to create a ClientSet: %v. Exiting.", err)
+			}
+			clients[i] = clientFromConfig
 		}
-		clients[i] = clientFromConfig
 	}
 
 	if config.Morph == "kubelet" {
 		f, c := kubemark.GetHollowKubeletConfig(config.createHollowKubeletOptions())
 
-		heartbeatClientConfigs := restclient.CopyConfigs(clientConfigs)
-		for _, heartbeatClientConfig := range heartbeatClientConfigs.GetAllConfigs() {
-			heartbeatClientConfig.Timeout = c.NodeStatusUpdateFrequency.Duration
-			// if the NodeLease feature is enabled, the timeout is the minimum of the lease duration and status update frequency
-			if utilfeature.DefaultFeatureGate.Enabled(features.NodeLease) {
-				leaseTimeout := time.Duration(c.NodeLeaseDurationSeconds) * time.Second
-				if heartbeatClientConfig.Timeout > leaseTimeout {
-					heartbeatClientConfig.Timeout = leaseTimeout
+		var heartbeatClient *clientset.Clientset
+		if enableInSecurePorts == false {
+			klog.V(2).Infof("create client config from file: %s", config.ResourceServer)
+			cfg, err := clientutil.CreateClientConfigFromKubeconfigFileAndSetQps(config.ResourceServer, 10, 20, config.ContentType)
```

Comment:
The previous QPS was set as -1, meaning using default QPS 5. Why we set here as 10? I am specially concern as this is from kubelet and could have big impact.

Reply:
Also we do check utilfeature.DefaultFeatureGate.Enabled(features.NodeLease) when enableInSecurePorts == true, why this is skipped here? I asked this question before but somehow it is not shown in the conversation.

Reply:
> Also we do check utilfeature.DefaultFeatureGate.Enabled(features.NodeLease) when enableInSecurePorts == true, why this is skipped here? I asked this question before but somehow it is not shown in the conversation.

Somehow I feel the if and else logic here can be consolidated.

```
var heartbeatClientConfig
var err error
if enableInSecurePorts == false {
    heartbeatClientConfig, err = clientutil.CreateClientConfigFromKubeconfigFileAndSetQps(config.ResourceServer, config.ContentType)
} else {
    heartbeatClientConfigs = restclient.CopyConfigs(clientConfigs)
}

for _, heartbeatClientConfig := range heartbeatClientConfigs.GetAllConfigs() {
    heartbeatClientConfig.Timeout = c.NodeStatusUpdateFrequency.Duration
    // if the NodeLease feature is enabled, the timeout is the minimum of the lease duration and status update frequency
    if utilfeature.DefaultFeatureGate.Enabled(features.NodeLease) {
        leaseTimeout := time.Duration(c.NodeLeaseDurationSeconds) * time.Second
        if heartbeatClientConfig.Timeout > leaseTimeout {
            heartbeatClientConfig.Timeout = leaseTimeout
        }
    }
    heartbeatClientConfig.QPS = float32(-1)
    if enableInSecurePorts == true && heartbeatClientConfig.Host != config.ResourceServer {
        heartbeatClientConfig.Host = config.ResourceServer
    }					
}

heartbeatClient, err = clientset.NewForConfig(heartbeatClientConfigs)
if heartbeatClientConfig.Host != config.ResourceServer {
    if err != nil {
        heartbeatClientConfig.Host = config.ResourceServer
        klog.Fatalf("Failed to create a ClientSet: %v. Exiting.", err)
    }
}

```

Reply:
looked back to the original code change in the PR backup branch prior to the refactor, this is a regression in the refactor code. it should be a call without the qps parameter which should be end up as the QPS not set to keep the default of them. will fix.

Reply:
fixed.